### PR TITLE
SONARJS-551 Rule: Strings and non-strings should not be added

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenatedWithNonStringCheck.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        if (stringLeft || stringRight) {
+          raiseIssue(tree);
+        }
+      }
+    }
+
+    super.visitBinaryExpression(tree);
+  }
+
+  private boolean isStringLiteral(Tree tree) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenatedWithNonStringCheck.java]: MAJOR Make "isStringLiteral" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/PrimitiveOperations.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+package org.sonar.javascript.tree.symbols.type;
+
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.sonar.plugins.javascript.api.symbols.Type;
+import org.sonar.plugins.javascript.api.symbols.Type.Kind;
+import org.sonar.plugins.javascript.api.tree.Tree;
+
+public class PrimitiveOperations {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/PrimitiveOperations.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1118)

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/TypeVisitor.java [hunk 3]
```diff
@@ -225,6 +227,16 @@ public void visitMemberExpression(MemberExpressionTree tree) {
     }
   }
 
+  @Override
+  public void visitBinaryExpression(BinaryExpressionTree tree) {
+    super.visitBinaryExpression(tree);
+
+    Type resultType = PrimitiveOperations.getType(tree.leftOperand().types().getUniqueKnownType(), tree.rightOperand().types().getUniqueKnownType(), ((JavaScriptTree) tree).getKind());
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/TypeVisitor.java]: MINOR Split this 184 characters long line (which is greater than 180 authorized). [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00103)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/PrimitiveOperations.java (hunk 2 above)
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/PrimitiveOperations.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1118)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenatedWithNonStringCheck.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  key = "S3402",
+  name = "Strings and non-strings should not be added",
+  priority = Priority.MAJOR,
+  tags = Tags.SUSPICIOUS)
+@ActivatedByDefault
+@SqaleSubCharacteristic(SubCharacteristics.INSTRUCTION_RELIABILITY)
+@SqaleConstantRemediation("15min")
+public class StringConcatenatedWithNonStringCheck extends BaseTreeVisitor {
+
+  private static final String MESSAGE = "Remove this addition.";
```
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenatedWithNonStringCheck.java]: @vilchik-elena It seems misleading to me to highlight a concatenation with a message like "Remove this addition."

Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenatedWithNonStringCheck.java]: @pynicolas do you have any suggestions?

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3402.html [hunk 5]
```diff
@@ -0,0 +1,26 @@
+<p>Use a <code>+</code> with two numbers and you'll get addition. But use it with a string and anything else, and you'll get concatenation. Very often that's not what's intended, and even when it is, it's likely to confuse future maintainers. </p>
+<p>Since this may not be what's intended, this rule raises an issue when <code>+</code> is used with a string and a non-string.</p>
+
+<h2>Noncompliant Code Example</h2>
+<pre>
+var x = 5 + 5;  // okay
+var z = "Hello" + 5;  // Noncompliant; yields: "Hello5". Was that intended?
```
Comment (pynicolas) [inline javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3402.html]: @vilchik-elena The implementation does not raise any issue for that line and for the following one.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenatedWithNonStringCheck.java (hunk 4 above)
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenatedWithNonStringCheck.java]: @vilchik-elena "Use the concat() method to make this concatenation explicit"? What if one of the operands is null?

Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenatedWithNonStringCheck.java]: @pynicolas actually we don't want user to use `concat` if concatenation is what he intends to do. But we want to catch cases when he intends to add but concatenate by mistake. WDYT about "Cast string operand "..." to number type."?

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3402.html (hunk 5 above)
Comment (vilchik-elena) [inline javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3402.html]: @pynicolas that's right! I forgot to update description. Fixed in 38273214be6c7d2c93b0de27658eb05e2f5cdd81

File: javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenatedWithNonStringCheck.java (hunk 4 above)
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenatedWithNonStringCheck.java]: @vilchik-elena What about a mix of the two? "Either make this concatenation explicit with the concat method or cast the left|right operand to a number"?

Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenatedWithNonStringCheck.java]: @pynicolas accepted

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/TypeVisitor.java [hunk 6]
```diff
@@ -197,6 +202,10 @@ public void visitIdentifier(IdentifierTree tree) {
     if (WebAPI.isDocument(tree)) {
       ((IdentifierTreeImpl) tree).addType(ObjectType.WebApiType.DOCUMENT);
     }
+
+    if (forLoopVariable) {
+      ((IdentifierTreeImpl) tree).addType(PrimitiveType.UNKNOWN);
```
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/TypeVisitor.java]: @vilchik-elena Why do we need that?

File: javascript-frontend/src/test/java/org/sonar/javascript/tree/symbols/type/ExpressionTypeTest.java [hunk 7]
```diff
@@ -66,6 +66,27 @@ public void new_expression() {
 
     par = getSymbol("o7");
     assertThat(par.types().containsOnlyAndUnique(Kind.DATE)).isTrue();
+  }
+
+  @Test
+  public void test_arithmetic_binary_operations() throws Exception {
```
Comment (pynicolas) [inline javascript-frontend/src/test/java/org/sonar/javascript/tree/symbols/type/ExpressionTypeTest.java]: @vilchik-elena shouldn't we test at least one case where we don't know the type of an operand?

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/TypeVisitor.java (hunk 6 above)
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/TypeVisitor.java]: @pynicolas to cover this test case https://github.com/SonarSource/sonar-javascript/pull/138/files#diff-d5c7ad5f7f5f1bab9ce946df5262fb31R84
We can have FP in 
```
var  x = 1;
var obj = {a: "a", b: "b"}
for (x of obj) {}  // we think that x is number, while here x is String
```

Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/type/TypeVisitor.java]: @vilchik-elena understood

File: javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenatedWithNonStringCheck.java (hunk 4 above)
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenatedWithNonStringCheck.java]: @vilchik-elena will you change it or do you prefer me to do it?

Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenatedWithNonStringCheck.java]: @pynicolas I will do that

Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenatedWithNonStringCheck.java]: @pynicolas fixed 9079d39e3732f1cc638d9c59d7873a54b46592c2

File: javascript-frontend/src/test/java/org/sonar/javascript/tree/symbols/type/ExpressionTypeTest.java (hunk 7 above)
Comment (vilchik-elena) [inline javascript-frontend/src/test/java/org/sonar/javascript/tree/symbols/type/ExpressionTypeTest.java]: @pynicolas fixed ffff2e7

**Pull request merged**